### PR TITLE
Enhancement/276/autodiscovery preset provider

### PR DESCRIPTION
**Pull request by cliffle**
Create the branch with a typo so all commits are now referenced to 275 instead of 276 :man_facepalming:

### CODE REVIEW

File: packages/sakuli-core/src/bootstrap/get-node-modules-paths.function.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  do {
+    const modules = join(path, "node_modules/@sakuli");
+    if (isExistingDirectory(modules)) {
+      nodeModulesPaths.push(modules);
+    }
+
+    previous = path;
+    path = resolve(join(previous, ".."));
+  } while (previous !== path);
```
Comment (svettwer) [inline packages/sakuli-core/src/bootstrap/get-node-modules-paths.function.ts:19]: Do I understand ist correctly that you scan every directory from `${workdir}` to `/` for `node_modules/@sakuli` subfolders? Why is this required? Wouldn't it be enough to scan the Sakuli project structure? We should not explore directories on a customers machine that don't belong to sakuli test suites.

In addition, the `@sakuli` search criteria is a strong restriction to presets developed by customers/users. They would have to publish their packages under `@sakuli` to leverage auto detection. They could still register it manually in `package.json` so it is not breaking the possibility to register custom presets but this is something to keep in mind.

File: packages/sakuli-core/src/bootstrap/get-sakuli-presets.function.spec.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+          "package.json": '{ "name": "just another package json" }',
+        },
+        "empty-dir": {},
+      },
+    });
+
+    const sakuliPresets = getSakuliPresets(["/base"]);
+
+    expect(sakuliPresets).toEqual(["sakuli"]);
+  });
```
Comment (svettwer) [inline packages/sakuli-core/src/bootstrap/get-sakuli-presets.function.spec.ts:22]: I find it pretty hard to follow the test case.
From the other parts of the code, in production, this code would process folder structures like `node_modules/@sakuli/someSakuliPresetProvider/package.json`, right?
What do you think about reflecting this within test case setup to better transport the use case of this function. I mean the function itself is abstract and can take/scan any location which is great, but it's hard to imagine what `/base/execa/package.json` tries to represent as such a structure is not common to the project.

File: packages/sakuli-core/src/bootstrap/get-node-modules-paths.function.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+  let { stdout } = await execa("npm", ["root", "-g"]);
```
Comment (svettwer) [inline packages/sakuli-core/src/bootstrap/get-node-modules-paths.function.ts:21]: I just realize you determine the root path here. I'd recommend to do this earlier in the code and use the following exit condition for directory scanning:

```
while(path !== npmRootPath)
```

**Comment by svettwer** [comment]
Hi @cliffle :wave: 
I just recognized the sonar gate failed even with the changes. I took a closer took and found a re-declaration of the require function. 
`declare function require(id: string): any;`
This was causing the issue. I'm not sure why, because it returns any but when I remove it, it passes. I've also tested if yargonaut still does its job to be sure not to break anything.

File: packages/sakuli-cli/src/commands/enable-enterprise/tasks/get-package-bootstrap-tasks.function.spec.ts [hunk 4]
```diff
@@ -1,42 +1,27 @@
...
-    })
-
-
-    it('should call proper task factories', () => {
-        const config = {};
-        const list = getPackageBootstrapTasks(
-            'package',
-            config
-        )
+    expect(list.length).toBe(2);
```
[the remark this answers is not in the document]
Comment (cliffle) [reply packages/sakuli-cli/src/commands/enable-enterprise/tasks/get-package-bootstrap-tasks.function.spec.ts:17]: The 3 tasks of the `sakuli enable-enterprise` were:
1. install the entered forwarder with sakuli
2. configure `sakuli.properties`
3. register the forwarder as preset in the `package.json`

With the preset autodiscovery, the third step is no longer necessary so I removed it. Therefore the task list reduced to 2.

File: packages/sakuli-core/src/bootstrap/get-node-modules-paths.function.ts (hunk 1 above)
Comment (cliffle) [reply packages/sakuli-core/src/bootstrap/get-node-modules-paths.function.ts]: Good point. I changed it so that only the global and the package level node_modules are searched.

I have limited this to `@sakuli` because there will be projects that could have hundreds of modules. So all modules would have to be searched for the `sakulipreset` keyword.

File: packages/sakuli-cli/src/commands/enable-enterprise/tasks/get-package-bootstrap-tasks.function.spec.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (svettwer) [reply packages/sakuli-cli/src/commands/enable-enterprise/tasks/get-package-bootstrap-tasks.function.spec.ts:17]: Ah, got it! It was about the number of tasks to execute for setup. :bulb: 
Thx!

File: packages/sakuli-core/src/bootstrap/get-node-modules-paths.function.ts (hunk 1 above)
Comment (svettwer) [reply packages/sakuli-core/src/bootstrap/get-node-modules-paths.function.ts]: A absolutely agree with the `@sakuli` limitation on automated search. Searching for presets in all `node_module` packages is the proverbial search for a needle in a haystack. It is just something, we have to keep in mind. In addition, I recognized manual loading was removed. Maybe we should keep it as a fallback solution for customer packages?

**Comment by cliffle** [comment]
Thanks for the review and the help on this PR :smile:
